### PR TITLE
[AMDGPU] Add target intrinsic for s_prefetch_data

### PR DESCRIPTION
**Pull request by rampitec**

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -19489,6 +19489,12 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
         F, {EmitScalarExpr(E->getArg(0)), EmitScalarExpr(E->getArg(1)),
             EmitScalarExpr(E->getArg(2)), EmitScalarExpr(E->getArg(3))});
   }
+  case AMDGPU::BI__builtin_amdgcn_s_prefetch_data: {
+    llvm::Value *Ptr = EmitScalarExpr(E->getArg(0));
```
Comment (shiltian) [inline clang/lib/CodeGen/CGBuiltin.cpp:19493]: can use `emitBuiltinWithOneOverloadedType<1>(...)`

Comment (rampitec) [reply clang/lib/CodeGen/CGBuiltin.cpp:19493]: I does not have a return value, so cannot.

Comment (shiltian) [reply clang/lib/CodeGen/CGBuiltin.cpp:19493]: `emitBuiltinWithOneOverloadedType` doesn't have to have a return value

Comment (rampitec) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Hm, ok, The description of the function is misleading. Done.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.prefetch.data.ll [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+; GCN-NEXT:    s_endpgm
+entry:
+  tail call void @llvm.amdgcn.s.prefetch.data.p4(ptr addrspace(4) %ptr, i32 %len)
+  ret void
+}
+
+define amdgpu_ps void @prefetch_data_sgpr_base_imm_len_global(ptr addrspace(1) inreg %ptr) {
+; GCN-LABEL: prefetch_data_sgpr_base_imm_len_global:
+; GCN:       ; %bb.0: ; %entry
+; GCN-NEXT:    s_prefetch_data s[0:1], 0x0, null, 31
```
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.prefetch.data.ll:96]: Why does this not have `v_readfirstlane_b32`? `isFlatGlobalAddrSpace` returns true for AS1 as well?

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.prefetch.data.ll:96]: That makes sense. Thanks for explanation.

**Review by shiltian — APPROVED** [review]

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 3]
```diff
@@ -2689,6 +2689,12 @@ def int_amdgcn_global_load_tr_b128 : AMDGPULoadIntrinsic<global_ptr_ty>;
 def int_amdgcn_wave_id :
   DefaultAttrsIntrinsic<[llvm_i32_ty], [], [NoUndef<RetIndex>, IntrNoMem, IntrSpeculatable]>;
 
+def int_amdgcn_s_prefetch_data :
+  Intrinsic<[], [llvm_anyptr_ty, llvm_i32_ty],
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2693]: Comment the operands? How is this different from the regular llvm.prefetch?

Comment (rampitec) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td]: Added comments. I will have to guard the regular llvm.prefetch, the instruction turns to be unsafe with invalid address. Besides target intrinsic can take length operand.

**Pull request merged**